### PR TITLE
feat(eks): Pass bootstrap.sh args to avoid DescribeCluster call and make nodes join the cluster faster

### PR DESCRIPTION
**Pull request by anguslees**
Pass sufficient information to bootstrap.sh in ASG user-data, to
prevent it from making a DescribeCluster AWS API call.  This per-node
API call can become an issue in large clusters, and we already have
all the information to avoid it.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-eks/lib/cluster.ts [hunk 1]
```diff
@@ -1632,6 +1632,16 @@ export interface BootstrapOptions {
...
 
+  /**
+
+   * Overrides the IP address to use for DNS queries within the
+   * cluster.
+   *
+   * @default - 10.100.0.10 or 172.20.0.10 based on the IP
+   * address of the primary interface.
+   */
+  readonly dnsClusterIp?: string;
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-eks/lib/cluster.ts:1643]: This doesn't seem relevant to the rest of the PR - can we split it up?

File: packages/@aws-cdk/aws-eks/lib/user-data.ts [hunk 2]
```diff
@@ -13,6 +13,8 @@ export function renderAmazonLinuxUserData(clusterName: string, autoScalingGroup:
 
   const extraArgs = new Array<string>();
 
+  extraArgs.push(`--apiserver-endpoint '${cluster.clusterEndpoint}'`);
+  extraArgs.push(`--b64-cluster-ca '${cluster.clusterCertificateAuthorityData}'`);
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-eks/lib/user-data.ts:17]: Can you reference some docs around these flags?

File: packages/@aws-cdk/aws-eks/lib/cluster.ts (hunk 1 above)
Comment (anguslees) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts:1643]: .. yep will do.  I was trying to sneak in support for this other bootstrap.sh flag at the same time because the per-change cost of contributing to cdk is so high :/  Will split it out..

File: packages/@aws-cdk/aws-eks/lib/user-data.ts (hunk 2 above)
Comment (anguslees) [reply packages/@aws-cdk/aws-eks/lib/user-data.ts:17]: What sort of docs are you looking for?  
Is just point to https://github.com/awslabs/amazon-eks-ami/blob/911119250416f460bbd60acd86e4763c006749fa/files/bootstrap.sh#L20-L21 sufficient ?

File: packages/@aws-cdk/aws-eks/lib/cluster.ts (hunk 1 above)
Comment (iliapolo) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts:1643]: Thanks.

Also - interested to hear about the costly parts of contributing, we are constantly looking for ways to improve that. If you drop you're feedback here i'll make sure it gets to the right place :)

File: packages/@aws-cdk/aws-eks/lib/user-data.ts (hunk 2 above)
Comment (iliapolo) [reply packages/@aws-cdk/aws-eks/lib/user-data.ts:17]: Yeap thats great. Just wanted to see somewhere those options are declared.

**Comment by iliapolo** [comment]
dont close

File: packages/@aws-cdk/aws-eks/lib/cluster.ts (hunk 1 above)
Comment (anguslees) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts:1643]: > interested to hear about the costly parts of contributing

For example, just ^here is ~3 weeks of PR grooming (a rebase/reupload about every week) before finding out that I need to make a trivial change (break out a single option into its own PR) before this PR will be accepted.

It's not _hard_, but this isn't my day job.  So checking/polling multiple outstanding PRs over a lengthy time period to see if there was some bitrot, or a reviewer has responded and the auto-close timer-of-doom has started is actually time consuming in aggregate.

(Normal community development stuff.  None of that is new or unique to this project, so I hope that doesn't sound too negative.)

Comment (iliapolo) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts:1643]: I understand. I agree that slow turn-around cycles are hurtful for productivity, and create contribution fatigue. Thanks for the feedback :pray:

File: packages/@aws-cdk/aws-eks/lib/user-data.ts [hunk 3]
```diff
@@ -3,7 +3,7 @@ import { Stack } from '@aws-cdk/core';
 import { BootstrapOptions, ICluster } from './cluster';
 
 // eslint-disable-next-line max-len
-export function renderAmazonLinuxUserData(clusterName: string, autoScalingGroup: autoscaling.AutoScalingGroup, options: BootstrapOptions = {}): string[] {
+export function renderAmazonLinuxUserData(cluster: ICluster, autoScalingGroup: autoscaling.AutoScalingGroup, options: BootstrapOptions = {}): string[] {
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-eks/lib/user-data.ts:6]: I just realized that for imported clusters, `cluster.clusterCertificateAuthorityData` might be undefined and will throw an exception.

Since this is only relevant for managed clusters at the moment, lets change to:

```suggestion
export function renderAmazonLinuxUserData(cluster: Cluster, autoScalingGroup: autoscaling.AutoScalingGroup, options: BootstrapOptions = {}): string[] {
```

**Comment by iliapolo** [comment]
@anguslees Are you still keeping an eye out on this PR?

File: packages/@aws-cdk/aws-eks/lib/user-data.ts (hunk 3 above)
Comment (anguslees) [reply packages/@aws-cdk/aws-eks/lib/user-data.ts:6]: I have changed to `Cluster`, as you suggested - but I will also note that `renderBottlerocketUserData` uses `cluster.clusterCertificateAuthorityData` from `ICluster`.

Future TODO: Ideally we would turn `clusterCertificateAuthorityData==undefined` into empty string (or not provide the bootstrap.sh arg in this case), and then bootstrap.sh will fall back to the old behaviour of querying at runtime - but it's currently an untyped exception :(

Comment (anguslees) [reply packages/@aws-cdk/aws-eks/lib/user-data.ts:6]: > Since this is only relevant for managed clusters at the moment, lets change to: ICluster->Cluster

.. No, wait.  Doesn't this need to work for imported clusters too?  I think `renderAmazonLinuxUserData` is used in building the worker node userdata.  And we definitely want to support adding worker nodes to clusters than came from `Cluster.fromClusterAttributes`.
 
(At least, I need to rewrite all the unittest fixtures for this function to stop using an imported `ICluster` - which I'm fine to do, but it made me stop and think.)

Comment (anguslees) [reply packages/@aws-cdk/aws-eks/lib/user-data.ts:6]: Ok, so I've changed this to catch the 'is not defined for this imported cluster' exception and silently skip the bootstrap.sh option in each case.  This will fallback to the existing behaviour (bootstrap.sh will discover these values by calling DescribeCluster at instance boot time).

The "ignore this exception" code is a moderately ugly string match - I tried replacing this with a typed exception, but extending `Error` seems to upset JSII.  Another option is to introduce new `ICluster` `tryClusterEndpoint()`, etc methods that return `undefined` rather than throw - that's a fairly large addition to a public class, but let me know if you'd prefer that version.

File: packages/@aws-cdk/aws-eks/lib/cluster.ts (hunk 1 above)
Comment (anguslees) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts]: Moved the dnsClusterIp parts to https://github.com/aws/aws-cdk/pull/13890
Note, the two PRs will trivially merge conflict in a test case, so one of them will need to be rebased and rereviewed.

File: packages/@aws-cdk/aws-eks/lib/user-data.ts (hunk 3 above)
Comment (iliapolo) [reply packages/@aws-cdk/aws-eks/lib/user-data.ts:6]: @anguslees 

> .. No, wait. Doesn't this need to work for imported clusters too? I think renderAmazonLinuxUserData is used in building the worker node userdata. And we definitely want to support adding worker nodes to clusters than came from Cluster.fromClusterAttributes

So...thats conceptually true, but notice that `user-data.ts`is not exported from `index.ts`, so its essentially a private API. And as far as our code goes, we only call it from `connectAutoScalingGroupCapacity`:

https://github.com/aws/aws-cdk/blob/567d64d70f92adbba9ff9981184d88b46fb95652/packages/%40aws-cdk/aws-eks/lib/cluster.ts#L1264-L1269

Which is only available on `Cluster`, not `ICluster`. I good with supporting `ICluster` if we can, but we should have an additional test to make sure we don't change the error message on one side only. 

> but I will also note that renderBottlerocketUserData uses cluster.clusterCertificateAuthorityData from ICluster

Yeap you're right, it should have been `Cluster` as well. Though I think that in this case, `cluster.clusterCertificateAuthorityData` has to be defined for the user-data to work, so while the context of the error won't be accurate, its ok for it to throw.

File: packages/@aws-cdk/aws-eks/test/test.user-data.ts [hunk 4]
```diff
@@ -155,12 +157,18 @@ function newFixtures(spot = false) {
   const app = new App();
   const stack = new Stack(app, 'my-stack', { env: { region: 'us-west-33' } });
   const vpc = new ec2.Vpc(stack, 'vpc');
+  const cluster = Cluster.fromClusterAttributes(stack, 'cluster', {
+    clusterName: 'my-cluster-name',
+    clusterCertificateAuthorityData: 'CADATA',
+    clusterEndpoint: 'https://mycluster:443',
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-eks/test/test.user-data.ts:163]: Let's add a test with an imported cluster that doesn't specify these properties, to make sure the exceptions are handled properly.

Comment (iliapolo) [inline packages/@aws-cdk/aws-eks/test/test.user-data.ts:163]: Lets add a test with an imported cluster that doesn't specify these properties, to make sure the exceptions are handled properly.

**Review by iliapolo — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/aws-eks/test/test.user-data.ts (hunk 4 above)
Comment (anguslees) [reply packages/@aws-cdk/aws-eks/test/test.user-data.ts:163]: > So...thats conceptually true, but notice that user-data.ts is not exported from index.ts, so its essentially a private API. And as far as our code goes, we only call it from connectAutoScalingGroupCapacity .. Which is only available on Cluster, not ICluster.

Ok .. so in that case I'm fine with assuming these properties are always defined, and I don't need to worry about catching/handling exceptions?

I think we went down this path _because_ we wanted to make it work with imported clusters (without raising exceptions) too.  I'm lost on what problem I'm trying to address here :P

Comment (iliapolo) [reply packages/@aws-cdk/aws-eks/test/test.user-data.ts:163]: > I'm lost on what problem I'm trying to address here :P

Ha yeah I think we got ourselves a little tangled here :)

> Ok .. so in that case I'm fine with assuming these properties are always defined, and I don't need to worry about catching/handling exceptions?

I would say that in that case it's better to just have this function accept `Cluster`, which will always have these properties defined. 

> I think we went down this path because we were concerned about backwards compatibility for imported clusters (ie: without raising exceptions)

Ah so yeah I guess its my bad, it wasn't about backwards compatibility ,I just didn't notice at the time that we can simply have this method accept `Cluster` instead of an `ICluster` - so I was trying to properly support that use case, with no need really.

So bottom line, lets just switch it to use `Cluster` :)

Comment (anguslees) [reply packages/@aws-cdk/aws-eks/test/test.user-data.ts]: Done.  The major change was to test.user-data.ts fixture/cases to avoid using ICluster.

**Review by iliapolo — DISMISSED** [review]

**Review by iliapolo — APPROVED** [review]
